### PR TITLE
Fuzz with cargo-fuzz

### PR DESCRIPTION
**Pull request by sanxiyn**

### CODE REVIEW

File: fuzz/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+target
+corpus
```
Comment (m4b) [inline fuzz/.gitignore]: Where is the corpus, don't we need a seed in order to fuzz? Or do you think we should let users add their own.

File: fuzz/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "goblin-fuzz"
+version = "0.0.1"
+authors = ["Automatically generated"]
```
Comment (m4b) [inline fuzz/Cargo.toml]: You can add yourself to authors here. I can make you Lead Liason and Executive Officer for Goblin Fuzzing if you like :)

File: fuzz/fuzz_targets/parse.rs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#![no_main]
+#[macro_use] extern crate libfuzzer_sys;
+extern crate goblin;
+
+fuzz_target!(|data: &[u8]| {
+    let _ = goblin::parse(data);
```
Comment (m4b) [inline fuzz/fuzz_targets/parse.rs]: Have you tried this? Last time iirc, it wouldn't work until I forced it to not pass 0 length bytes.

File: fuzz/.gitignore (hunk 1 above)
Comment (sanxiyn) [inline fuzz/.gitignore]: cargo-fuzz can fuzz without corpus.

File: fuzz/fuzz_targets/parse.rs (hunk 3 above)
Comment (sanxiyn) [inline fuzz/fuzz_targets/parse.rs]: Yes, I have tried this, and it works. It finds e_phoff overflow I already reported, and also very similar e_shoff overflow.

File: fuzz/Cargo.toml (hunk 2 above)
Comment (sanxiyn) [inline fuzz/Cargo.toml]: Done!

File: fuzz/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+publish = false
+
+[package.metadata]
+cargo-fuzz = true
+
+[dependencies.goblin]
+path = ".."
+
+[dependencies.libfuzzer-sys]
+git = "https://github.com/rust-fuzz/libfuzzer-sys.git"
```
Comment (m4b) [inline fuzz/Cargo.toml]: Does this have to be a git dependency?

Comment (sanxiyn) [inline fuzz/Cargo.toml]: Unfortunately, yes, for now.

**Comment by m4b** [comment]
Wooohooooo thanks for this! :)

**Comment by m4b** [comment]
@sanxiyn 

Actually how do you run this ? :laughing: 

I get:

```
m4b@efrit ::  [ ~/projects/goblin/fuzz ] cargo run
    Updating git repository `https://github.com/rust-fuzz/libfuzzer-sys.git`
    Updating registry `https://github.com/rust-lang/crates.io-index`
 Downloading arbitrary v0.1.0
   Compiling unicode-xid v0.0.4
   Compiling gcc v0.3.51
   Compiling scroll v0.5.0
   Compiling quote v0.3.15
   Compiling arbitrary v0.1.0
   Compiling plain v0.0.2
   Compiling synom v0.11.3
   Compiling syn v0.11.11
   Compiling libfuzzer-sys v0.1.0 (https://github.com/rust-fuzz/libfuzzer-sys.git#67f73995)
   Compiling scroll_derive v0.4.0
   Compiling goblin v0.0.10 (file:///home/m4b/projects/goblin)
   Compiling goblin-fuzz v0.0.1 (file:///home/m4b/projects/goblin/fuzz)
    Finished dev [unoptimized + debuginfo] target(s) in 19.71 secs
     Running `target/debug/parse`
WARNING: Failed to find function "__sanitizer_get_total_unique_coverage".
WARNING: Failed to find function "__sanitizer_print_stack_trace".
WARNING: Failed to find function "__sanitizer_reset_coverage".
WARNING: Failed to find function "__sanitizer_set_death_callback".
INFO: Seed: 269520155
ERROR: __sanitizer_set_death_callback is not defined. Exiting.
Did you use -fsanitize-coverage=... to build your code?
```

**Comment by sanxiyn** [comment]
https://github.com/rust-fuzz/cargo-fuzz explains how. For example,

```
cargo fuzz run parse -- -seed=1
cargo fuzz run parse -- -seed=2
RUST_BACKTRACE=1 cargo fuzz run parse fuzz/artifacts/parse/crash-7* # e_phoff overflow
RUST_BACKTRACE=1 cargo fuzz run parse fuzz/artifacts/parse/crash-0* # e_shoff overflow
```

**Comment by m4b** [comment]
Awesome, thanks! 

So if you want to try the `scroll_no_offset` branch I just pushed, for me it's been running for a while but i don't seem to have any errors so far

**Comment by m4b** [comment]
Actually as I wrote this I oom'd, but I knew about these before, with large values in the dynamic array or header; only correct way to do this is to check the length requested with the size of the binary, only for some types (e.g., bss this won't work)

**Comment by sanxiyn** [comment]
I tried scroll_no_offset branch, and it is indeed much better!

With seed 4, I get another overflow at src/strtab.rs:37. (To reproduce, be sure to delete fuzz/corpus.)

**Comment by sanxiyn** [comment]
Seed 10 gives index out of bounds at src/elf/mod.rs:205. (Corrupt sh_link.)

**Pull request merged**